### PR TITLE
libxml2: use libiconv on Darwin

### PR DESCRIPTION
**Pull request by emilazy**
###### Motivation for this change

This fixes e.g. `python3Packages.beautifulsoup4`, which has tests relying on the wider encoding support enabled by `libiconv`. Not built locally because I don't have the resources to bootstrap stdenv, but I made `python3Packages.lxml` use this version of `libxml2` and everything seemed to work fine. If it's not acceptable to include `libiconv` in the Darwin stdenv bootstrap, let me know and I can see about conditionalizing it.

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by emilazy** [comment]
One alternative would be to continue to omit `libiconv` and just disable the failing encoding-related BeautifulSoup tests.

**Comment by risicle** [comment]
Trying to reproduce your `lxml`-building and getting downstream packages themselves wanting iconv... e.g.

```
 /nix/store/mbqgykviliwg19bfajsz79bfbg344mlq-libxml2-2.9.12-dev/include/libxml2/libxml/encoding.h:28:10: fatal error: 'iconv.h' file not found
#include <iconv.h>
         ^~~~~~~~~
1 error generated.
```

When trying to build `libxslt`.

**Comment by risicle** [comment]
This is on macos 10.15. `libiconv` doesn't seem to get included in `nix-support/propagated-build-inputs`

**Comment by emilazy** [comment]
I think what's going on here is that the aarch64-darwin SDK (which is based on a newer macOS SDK than the x86_64-darwin one) has an `iconv.h` header but no `iconv` library, so (a) this header issue "just works out" on aarch64, and (b) there's a bunch of other weird downstream issues of build systems detecting `iconv.h` and expecting an `iconv` library that isn't there.

I'm not sure what the correct resolution is; is `propagatedBuildInputs` usually used for header dependencies like this?

File: pkgs/top-level/all-packages.nix [hunk 1]
```diff
@@ -17759,6 +17759,7 @@ with pkgs;
 
   libxml2 = callPackage ../development/libraries/libxml2 {
     python = python3;
+    inherit (darwin) libiconv;
```
Comment (SuperSandro2000) [inline pkgs/top-level/all-packages.nix:17762]:
```suggestion
```

**Comment by emilazy** [comment]
Uses `propagatedBuildInputs` now, which should hopefully fix the build issue on `x86_64-darwin`.

File: pkgs/top-level/all-packages.nix (hunk 1 above)
Comment (emilazy) [reply pkgs/top-level/all-packages.nix:17762]: This is fairly commonly used in `all-packages.nix` for cases where the dependency is only used on Darwin, I figured it's a little clearer this way rather than relying on the conditional/stubbed-out package (iconv is part of glibc). Maybe the other cases should be changed too though? I'm not aware of any particular standard one way or the other for this.

Comment (r-burns) [reply pkgs/top-level/all-packages.nix:17762]: I noticed a couple days ago, that we should maybe update the definition of libiconv: https://github.com/NixOS/nixpkgs/pull/138545#discussion_r713755466

Maybe something we can tackle in a separate PR

Comment (emilazy) [reply pkgs/top-level/all-packages.nix:17762]: Hm, I'm a little confused; shouldn't `!stdenv.isLinux` imply `stdenv.hostPlatform.libc != "glibc"` in almost all cases to begin with? What's incorrect about the current `libiconv`/`libintl` definitions?

**Review by risicle — APPROVED** [review]
This works for me now, macos 10.15 (and manages to build `beautifulsoup4`). Still wish we could figure out why it's not being picked up as it should.

**Comment by risicle** [comment]
Might it be possible to get this in `staging-next` given than the breakage is currently preventing thousands of packages being properly reviewed against darwin?

**Comment by emilazy** [comment]
Going by ofborg, it seems like this would break `stdenv`:

```
building '/nix/store/7h383zwwz8k5vjr52qc0j1yjcqc39pig-bootstrap-stage3-stdenv-darwin.drv'...
output '/nix/store/jswbz9pynnfcn96p3j2q70j0x3s128nf-bootstrap-stage3-stdenv-darwin' is not allowed to refer to the following paths:
  /nix/store/dhkh15znvls5s4rnqk71pcmrrs95r7x9-libiconv-50
```

Either the restriction needs loosening (and I'm not sure how), or we need a `libxml2Minimal` to use in stdenv bootstrap, or we should go with an alternate approach like selectively disabling the Beautiful Soup XML encoding tests on Darwin.

**Comment by SuperSandro2000** [comment]
> Either the restriction needs loosening (and I'm not sure how)

see https://github.com/NixOS/nixpkgs/pull/115498/files

**Comment by emilazy** [comment]
Thanks. Strangely it was already present in most of those locations, but I added it to the remaining ones and hopefully that'll fix things.

**Comment by emilazy** [comment]
I'm not sure why this is still failing. `libiconv` is present in all the lists `libxml2` is in the Darwin stdenv file, as far as I can tell. Is it something about needing `.out`, maybe?

**Comment by risicle** [comment]
Afraid was only able to test through hackiness to avoid the bootstrap loop :(

**Comment by risicle** [comment]
Would it be helpful if I prepared a PR aimed at `staging-next` skipping the bs4 test in question, just to get darwin's python ecosystem back sooner rather than later? (it's not like I'd be leaving anything more broken than it was before, right?)

Or might reliving the pressure in that way also relieve the pressure to get it fixed properly?

**Comment by happysalada** [comment]
It looks like the ofborg failure was with stage3 and it seems you added the file to the persistent list of stage 2.
I'm not sure why it's failing like this, since libiconv is in the `allowedRequisites` of stage3. I had to modify this part once, and for me just adding the missing derivation to allowed requisites fixed that particular error.
I should have some time to test this further tonight.

**Comment by happysalada** [comment]
I tried building bash and pcre on that PR and it all went fine. I didn't see the failure before with stage3.
I'm going to test quickly without adding libiconv to stage2

**Comment by happysalada** [comment]
Removing that from stage2 makes it fail.
I've tried modifying the allowedRequisites to add the `.out` at the end, but no matter what I do, I couldn't get it to work.
It looks that stage2 reference is needed.

**Comment by happysalada** [comment]
The latest ofborg failure https://github.com/NixOS/nixpkgs/pull/137870/checks?check_run_id=3748738841 just says time out.
I'm going to go ahead and merge this, since it looks correct.

**Comment by happysalada** [comment]
Unless I'm wrong, I think this update breaks asciidoctor on darwin.
asciidoctor relies on nokogiri that itself relies on libxml2.
Upon compilation of nokogiri, the following errors occurs
```
In file included from /nix/store/1vpjj9aza91r164sfmcyjjnjfnczlps5-libxml2-2.9.12-dev/include/libxml2/libxm
l/parser.h:812:
/nix/store/1vpjj9aza91r164sfmcyjjnjfnczlps5-libxml2-2.9.12-dev/include/libxml2/libxml/encoding.h:28:10: fa
tal error: 'iconv.h' file not found
#include <iconv.h>
```

Nokogiri itself is including libiconv, however the flag passed to the compilation for nokogiri is not being taken in account when looking for the header in the libxml2 file.
Here is where nokogiri itself requires iconv https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/ruby-modules/gem-config/default.nix#L440

It took me a while to find the fix, so I'm just documenting here in case something similar happens to another dependent of libxml2
Basically you have to add a flag to tell it where to look for additional header.
See https://github.com/NixOS/nixpkgs/pull/144582
